File: leetcode-practice-java/src/main/java/io/hank/leetcode/practices/twopointer/_16_3Sum_Closest.java
package io.hank.leetcode.practices.twopointer;

import io.hank.leetcode.annotations.*;
import io.hank.leetcode.practices.LeetcodeProblemSolution;

import java.util.Arrays;

/**
 * <pre>
 * Given an integer array nums of length n and an integer target, find three integers in nums such that the sum is closest to target.
 *
 * Return the sum of the three integers.
 *
 * You may assume that each input would have exactly one solution.
 *
 *
 *
 * Example 1:
 *
 * Input: nums = [-1,2,1,-4], target = 1
 * Output: 2
 * Explanation: The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).
 *
 * Example 2:
 *
 * Input: nums = [0,0,0], target = 1
 * Output: 0
 * Explanation: The sum that is closest to the target is 0. (0 + 0 + 0 = 0).
 *
 *
 * Constraints:
 *
 * 3 <= nums.length <= 500
 * -1000 <= nums[i] <= 1000
 * -10^4 <= target <= 10^4
 * </pre>
 */
public class _16_3Sum_Closest extends LeetcodeProblemSolution {

    @Topic(TopicType.TWO_POINTER)
    @TimeComplexity(ComplexityType.O_N2)
    @SpaceComplexity(ComplexityType.O_1)
    int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int result = nums[0] + nums[1] + nums[nums.length - 1];
        if (result == target) {
            return target;
        }
        int delta = Math.abs(target - result);

        for (int i = 0; i < nums.length - 2; i++) {
            int j = i + 1;
            int k = nums.length - 1;
            int newDelta = 0;
            while (j < k) {
                int sum = nums[i] + nums[j] + nums[k];
                if (sum == target) {
                    return target;
                } else if (sum < target) {
                    j++;
                    newDelta = target - sum;
                } else {
                    k--;
                    newDelta = sum - target;
                }

                if (newDelta < delta) {
                    result = sum;
                    delta = newDelta;
                }
            }
        }

        return result;
    }
}

File: leetcode-practice-python/src/two_pointers/16_3sum_closest.py
class Solution:
    def threeSumClosest(self, nums: list[int], target: int) -> int:
        nums.sort()
        n = len(nums)
        a, b, c = nums[0], nums[1], nums[n - 1]
        min_delta = abs(a + b + c - target)
        ans = a + b + c
        if min_delta == 0:
            return target

        # 迭代每个a的可能值
        for i in range(0, n - 2):
            a = nums[i]
            # 迭代每个b和c的可能
            j, k = i + 1, n - 1
            while j < k:
                b, c = nums[j], nums[k]
                sum_of_abc = a + b + c
                new_diff = abs(sum_of_abc - target)
                if sum_of_abc > target:
                    k -= 1
                elif sum_of_abc == target:
                    return target
                else:
                    j += 1
                if new_diff < min_delta:
                    min_delta = new_diff
                    ans = sum_of_abc

        return ans


if __name__ == '__main__':
    s = Solution()
    print(s.threeSumClosest([-1, 2, 1, -4], 1))
    print(s.threeSumClosest([0, 0, 0], 1))
    print(s.threeSumClosest([1, 1, 1, 0], -100))
